src/main/java/edu/harvard/iq/dataverse/AuxiliaryFileServiceBean.java

```diff
@@ -0,0 +1,114 @@
+
+package edu.harvard.iq.dataverse;
+
+import edu.harvard.iq.dataverse.dataaccess.StorageIO;
+import edu.harvard.iq.dataverse.util.FileUtil;
+import edu.harvard.iq.dataverse.util.SystemConfig;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.logging.Logger;
+import javax.ejb.EJB;
+import javax.ejb.Stateless;
+import javax.inject.Named;
+import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.Query;
+import org.apache.tika.Tika;
+
+/**
+ *
+ * @author ekraffmiller
+ *  Methods related to the AuxiliaryFile Entity.
+ */
+@Stateless
+@Named
+public class AuxiliaryFileServiceBean implements java.io.Serializable {
+   private static final Logger logger = Logger.getLogger(AuxiliaryFileServiceBean.class.getCanonicalName());
+
+    @PersistenceContext(unitName = "VDCNet-ejbPU")
+    private EntityManager em;
+    
+    @EJB
+    private SystemConfig systemConfig;
+    
+
+    public AuxiliaryFile find(Object pk) {
+        return em.find(AuxiliaryFile.class, pk);
+    }
+
+    public AuxiliaryFile save(AuxiliaryFile auxiliaryFile) {
+        AuxiliaryFile savedFile = em.merge(auxiliaryFile);
+        return savedFile;
+
+    }
+    
+    /**
+     * Save the physical file to storageIO, and save the AuxiliaryFile entity
+     * to the database.  This should be an all or nothing transaction - if either
+     * process fails, than nothing will be saved
+     * @param fileInputStream - auxiliary file data to be saved
+     * @param dataFile  - the dataFile entity this will be added to
+     * @param formatTag - type of file being saved
+     * @param formatVersion - to distinguish between multiple versions of a file
+     * @param origin - name of the tool/system that created the file
+     * @param isPublic boolean - is this file available to any user?
+     * @return success boolean - returns whether the save was successful
+     */
+    public boolean processAuxiliaryFile(InputStream fileInputStream, DataFile dataFile, String formatTag, String formatVersion, String origin, boolean isPublic) {
+    
+        StorageIO<DataFile> storageIO =null;
+      
+        String auxExtension = formatTag + "_" + formatVersion;
+        try {
+            // Save to storage first.
+            // If that is successful (does not throw exception),
+            // then save to db.
+            // If the db fails for any reason, then rollback
+            // by removing the auxfile from storage.
+            storageIO = dataFile.getStorageIO();
+            AuxiliaryFile auxFile = new AuxiliaryFile();
+            storageIO.saveInputStreamAsAux(fileInputStream, auxExtension);
+            auxFile.setChecksum(FileUtil.calculateChecksum(storageIO.getAuxFileAsInputStream(auxExtension), systemConfig.getFileFixityChecksumAlgorithm()));
```

Comment:
It's not too hard to wrap the fileInputStream and calculate the has during the original write (versus having to retrieve the file and scan it a second time. (The util.Bag code has examples.)

Reply:
Can you explain more what you had it mind? I looked at BagGenerator and BagValidation map, and can't find an example of this.  I could see how you could combine the logic that reads the stream to calculate the checksum, with the read from stream that happens in StorageIO, but I don't see an example of where that's happening.

Reply:
Sorry - it would help if I pointed to the right place. It's actually the archiving commands that do this. E.g. [the Google Cloud one](https://github.com/IQSS/dataverse/blob/develop/src/main/java/edu/harvard/iq/dataverse/engine/command/impl/GoogleCloudSubmitToArchiveCommand.java#L78 ). They complicate things with threads, but the basic idea is that you wrap your stream with a java.security.DigestInputStream and, once you've finished reading the bytes, you can get the digest for the algorithm you set from that stream.

Reply:
ok, that makes sense, thank you